File: Lab 8/DataStructures/SinglyLinkedList.cs
using System;

namespace DataStructures
{
    public class SinglyLinkedList
    {
        public Node First { get; private set; }
        public Node Last { get; private set; }
        public int Size { get; private set; }
        public void Add(short value)
        {
            if (First != null)
            {
                Node node = new Node(value);
                node.Next = First;
                First = node;
                Size++;
            }
            else
            {
                First = Last = new Node(value);
                Size = 1;
            }
        }
        public int QuantityMultiplesOfThree()
        {
            int counter = 0;
            Node currentNode = First;
            while (currentNode != null)
            {
                if (currentNode.Value % 3 == 0)
                {
                    counter++;
                }
                currentNode = currentNode.Next;
            }
            return counter;
        }
        public void RemoveLargeForAverage()
        {
            double mean = 0;
            Node currentNode = First;
            while (currentNode != null)
            {
                mean += currentNode.Value;
                currentNode = currentNode.Next;
            }
            mean /= Size;
            Node prevNode = null;
            currentNode = First;
            while (currentNode != null)
            {
                if (currentNode.Value > mean)
                {
                    if (prevNode == null)
                    {
                        First = currentNode.Next;
                        Size--;
                    }
                    else
                    {
                        prevNode.Next = currentNode.Next;
                        Size--;
                    }
                }
                else
                {
                    prevNode = currentNode;
                }
                currentNode = currentNode.Next;
            }
            Last = prevNode;
        }
        public short this[int index]
        {
            get
            {
                if (index < 0 || index >= Size)
                {
                    throw new IndexOutOfRangeException("Index out of range");
                }
                if (Size == 0)
                {
                    throw new InvalidOperationException("The list is empty");
                }
                int counter = 0;
                short value = default;
                bool find = false;
                Node currentNode = First;
                while (currentNode != null&&!find)
                {
                    if (counter == index)
                    {
                        value = currentNode.Value;
                        find = true;
                    }
                    currentNode = currentNode.Next;
                    counter++;
                }
                return value;
            }
        }
    }
}

File: Lab 8/C++/SinglyLinkedList.cpp
#include "SinglyLinkedList.h"
#include <stdexcept>
#include <exception>

using namespace std;
Node::Node(short value)
{
	this->value = value;
    this->next = nullptr;
}

void Node::SetNext(Node* next)
{
    this->next = next;
}

Node* Node::GetNext()
{
    return next;
}

short Node::GetValue()
{
    return value;
}

SinglyLinkedList::SinglyLinkedList()
{
    size = 0;
    first = nullptr;
    last = nullptr;
}

void SinglyLinkedList::Add(short value)
{
    if (first != nullptr)
    {
        Node* node = new Node(value);
        node->SetNext(first);
        first = node;
        size++;
    }
    else
    {
        first = last = new Node(value);
        size = 1;
    }
}

int SinglyLinkedList::QuantityMultiplesOfThree()
{
    int counter = 0;
    Node* currentNode = first;
    while (currentNode != nullptr)
    {
        if (currentNode->GetValue() % 3 == 0)
        {
            counter++;
        }
        currentNode = currentNode->GetNext();
    }
    return counter;
}

void SinglyLinkedList::RemoveLargeForAverage()
{
    double mean = 0;
    Node* currentNode = first;
    while (currentNode != nullptr)
    {
        mean += currentNode->GetValue();
        currentNode = currentNode->GetNext();
    }
    mean /= size;
    Node* prevNode = nullptr;
    currentNode = first;
    while (currentNode != nullptr)
    {
        if (currentNode->GetValue() > mean)
        {
            if (prevNode == nullptr)
            {
                first = currentNode->GetNext();
                size--;
            }
            else
            {
                prevNode->SetNext(currentNode->GetNext());
                size--;
            }
        }
        else
        {
            prevNode = currentNode;
        }
        currentNode = currentNode->GetNext();
    }
    last = prevNode;
}

short SinglyLinkedList::operator[](const int index)
{
    if (index < 0 || index >= size)
    {
        throw out_of_range("Index out of range");
    }
    if (size == 0)
    {
        throw logic_error("The list is empty");
    }
    int counter = 0;
    Node* currentNode = first;
    while (currentNode != nullptr)
    {
        if (counter == index)
        {
            return currentNode->GetValue();
        }
        currentNode = currentNode->GetNext();
        counter++;
    }
}

int SinglyLinkedList::Size()
{
    return size;
}
